### PR TITLE
fix(lsp): decoding percent-encoding(non-ASCII) file path correctly

### LINKED ISSUE #12065: lsp: consider handling unicode better
From denoland/vscode_deno#517 by @TradeIdeasPhilip:

> _intellisense_
> 
> 1. Create a file with a name like `¡Feliz cumpleaños!.ts` or `⛔message-from-server.ts`.
> 2. In a different file, reference an item that is exported from the file with the non-English characters in the name.
> 3. Intellisense will suggest adding an import.  Accept its suggestion.
> 4. The line it creates will quote all of the non-English characters like this:  `import { celebrate1 } from "./%C2%A1Feliz%20cumplea%C3%B1os!.ts";`
> 5. You can keep that ugly file name with the URL-style quotes or you can change it to the easy to read format.  It doesn't make a difference.
> 6. Either way the program will run fine.
> 7. Either way the debugger will act up like described above.

### PR DESCRIPTION
**Pull request by Hajime-san**
# Summary

The `Url` alias of [ModuleSpecifier](https://github.com/denoland/deno_core/blob/0a3cc0509ded8b621330698a2dd32c6a5121bc73/core/module_specifier.rs#L52) type in `deno_core` keeps local file path with percent-encoding string while it includes non-ASCII character like emoji.
It causes unintentional completion, hover text, rename and diagnostic with percent-encoding string by lsp server.

This patch is an ad hoc approach for resolve this problem.

## Expected behaviour

https://github.com/denoland/deno/assets/41257923/9ca279fc-46d6-4189-bfdf-3f48685946e7

## related issues

resolves #12065
resolves https://github.com/denoland/vscode_deno/issues/517

There are test reporter and runtime problems.
#18983

There is a test reporter problem.
#10810

There are tsc problems.
#16058
https://github.com/microsoft/TypeScript/issues/57539

### CODE REVIEW

File: cli/lsp/documents.rs [hunk 1]
```diff
@@ -647,15 +648,20 @@ pub fn to_hover_text(result: &Resolution) -> String {
   match result {
     Resolution::Ok(resolved) => {
       let specifier = &resolved.specifier;
+      let format = |scheme: &str, rest: &str| -> String {
+        format!("{}&#8203;{}", scheme, rest).replace('@', "&#8203;@")
+      };
```
Comment (Hajime-san) [inline cli/lsp/documents.rs:653]: BTW, I'm not clear to put in zero-width space👀

**Comment by Hajime-san** [comment]
Oh, `jsr` is invalid about unicode file path...🦕

- https://jsr.io/docs/troubleshooting#invalidpath

**Review by aapoalas — DISMISSED** [review]
Conflict with main: `to_hover_text` has been removed in main apparently?

**Comment by Hajime-san** [comment]
Fix with rebase!

File: cli/lsp/tsc.rs [hunk 2]
```diff
@@ -547,6 +548,10 @@ impl TsServer {
       .and_then(|mut changes| {
         for changes in &mut changes {
           changes.normalize(&self.specifier_map)?;
+          for text_chenages in &mut changes.text_changes {
+            text_chenages.new_text =
+              to_percent_decoded_str(&text_chenages.new_text);
```
Comment (nayeemrmn) [inline cli/lsp/tsc.rs:553]:
```suggestion
          for text_changes in &mut changes.text_changes {
            text_changes.new_text =
              to_percent_decoded_str(&text_changes.new_text);
```

**Review by nayeemrmn — APPROVED** [review]
Thanks! Essentially this patch formats 'pretty' percent-decoded URLs for the purpose of error messages, auto-import and auto-renamed specifiers. But these URLs will normalise to the same thing.

File: cli/lsp/tsc.rs (hunk 2 above)
Comment (Hajime-san) [reply cli/lsp/tsc.rs]: Thanks! Fix this.

**Comment by Hajime-san** [comment]
Thanks!
If this PR merged, I'll tackle to fix test reporter path correctly with the `to_percent_decoded_str` function in  `cli/util/path.rs`

**Pull request merged**